### ISSUE TITLE
[HELP] How to implement ota on tb-gateway?

### ISSUE THREAD

**Issue by lincoln310** [thingsboard/thingsboard-gateway#953]
**Describe the issue**
How to implement ota on tb-gateway? Do you have this play? What should i do if i want to add it?

**Versions (please complete the following information):**
 - OS: [e.g. Ubuntu 20.04]
 - Thingsboard IoT Gateway version [e.g. 3.4.2]
 - Python version[e.g. 3.9]

**Comment by samson0v** [comment]
Hi @lincoln310, for OTA API you can use our [Python SDK](https://github.com/thingsboard/thingsboard-python-client-sdk).

**Comment by lincoln310** [comment]
@samson0v Thanks! But what i want is : my device connect to tb-gateway, and get firmware from it. So the tb-gateway should work with ota message as other attribute message. Like a bridge.

**Comment by samson0v** [comment]
@lincoln310, Gateway doesn't support OTA API for now.

**Comment by lincoln310** [comment]
> @lincoln310, Gateway doesn't support OTA API for now.

Indeed, so what should i do to add it? Any suggestion? I can send a pull request when i finish.

**Comment by imbeacon** [comment]
Hi @lincoln310,

Unforunately at the moment ThingsBoard doesn't have API to get firmware for devices through the gateway, we will discuss this point with team and let you know about the results.

**Comment by lincoln310** [comment]
@imbeacon I need 2 function:
1  ota for my device. This should connect to tb or tb-edge
2  custom binary data process on custom topic from device. This should use tb-Gateway

Any suggestion?

**Comment by imbeacon** [comment]
@lincoln310 ,

1. There is no API on ThenigsBoard for the gateway to get firmware for devices connected through it, so there is no way at the moment to implement this on the gateway by now, unfortunatly.
2. We are going to add different types for default mqtt converter like string, bytes in the nearest future, we will write here when add them.

**Comment by lincoln310** [comment]
I can get latest attribute about firmware on device that connected to gw, when update firmware on web.
So maybe i can just impletment in tb-gateway like this:
subscribe topic v2/fw/request/${deviceName}/chunk/${chunkIdx}
and send http request to tb to get the response, 
and then send as mqtt msg to v2/fw/response/${deviceName}/chunk/${chunkIdx}
and then combind the chunk data.

How about it?

**Comment by arthurgustin** [comment]
Any news on this topic ? I try to implement OTA with the gateway too

**Comment by lincoln310** [comment]
> Any news on this topic ? I try to implement OTA with the gateway too

Sorry, i give it up.

**Comment by miraclebg** [comment]
I've already done this and it works - with a custom firmware connector and some small changes to the MQTT connector.

When i have some time i'll create a pull request to review.

If someone needs it desperately ping me at miracle[at]nimasystems.com.

**Comment by xianyo** [comment]
I need this feature; I’m looking forward to it. @miraclebg

**Comment by samson0v** [comment]
@miraclebg @xianyo @lincoln310 @arthurgustin 📢 a little announcement:
We just launched our [official Discord server](https://discord.com/invite/mJxDjAM3PF), we'd love to see you there!

**Comment by apachler** [comment]
@miraclebg this would be a cool feature!
